packages/flutter/lib/src/cupertino/date_picker.dart

```diff
@@ -1026,22 +1086,48 @@ class _CupertinoTimerPickerState extends State<CupertinoTimerPicker> {
   Alignment alignCenterLeft;
   Alignment alignCenterRight;
 
+  FixedExtentScrollController hourController;
+  FixedExtentScrollController minuteController;
+  FixedExtentScrollController secondController;
+
   // The currently selected values of the picker.
   int selectedHour;
   int selectedMinute;
   int selectedSecond;
 
+  CupertinoTimerPickerController timerController;
+
   @override
   void initState() {
     super.initState();
 
-    selectedMinute = widget.initialTimerDuration.inMinutes % 60;
+    timerController = widget.controller ?? CupertinoTimerPickerController();
+    timerController.addListener(_desiredDurationChanged);
+
+    // TODO(jslavitz): Remove this check and fully deprecate widget.initialTimerDuration
+    final Duration initialDuration =
+      widget.initialTimerDuration != Duration.zero ? // ignore: deprecated_member_use_from_same_package
+      widget.initialTimerDuration : widget.controller.desiredTimerDuration; // ignore: deprecated_member_use_from_same_package
+
+    selectedMinute = initialDuration.inMinutes % 60;
 
     if (widget.mode != CupertinoTimerPickerMode.ms)
-      selectedHour = widget.initialTimerDuration.inHours;
+      selectedHour = initialDuration.inHours;
 
     if (widget.mode != CupertinoTimerPickerMode.hm)
-      selectedSecond = widget.initialTimerDuration.inSeconds % 60;
+      selectedSecond = initialDuration.inSeconds % 60;
+
+    timerController = widget.controller ?? CupertinoTimerPickerController();
+
+    hourController = FixedExtentScrollController(initialItem: selectedHour);
+    minuteController = FixedExtentScrollController(initialItem: selectedMinute ~/ widget.minuteInterval);
+    secondController = FixedExtentScrollController(initialItem: selectedSecond ~/ widget.secondInterval);
+  }
+
+  @override
+  void didUpdateWidget(CupertinoTimerPicker oldWidget) {
+    timerController ??= widget.controller;
```

Comment:
Do you need to remove the listener from an existing controller, and add a listener to the new one?

If I'm right about that then we should add some tests in this PR that covers controller edge cases because this stuff is easy to miss.

Reply:
True true true!

Reply:
Yup, will do!